### PR TITLE
Adding tests for `Remove-Module`

### PR DESCRIPTION
**Pull request by pougetat**
# PR Summary

This PR adds approx 50 tests (+3 seconds when running locally) for the remove-module  cmdlet. 

If anyone has suggestions for test cases I missed please feel free to comment. Some of the cases listed are quite vague for the moment. My objective is to optimize the ````number of tests written / branches tested````. I will also screenshot some codeCov results once I get around to running it on the codebase.

Tests have been added tests for : 
  - [x] Remove-Module -Name
      - [x] simple module name (single, multiple, repeated)
      - [x] regular expressions (single, multiple, no matches) ["*" case breaks Pester (see code comment)]
      - [x] error cases (no matches for simple names)
  - [x] Remove-Module -FullyQualifiedName
     - [x] using ModuleVersion
     - [x] using RequiredVersion
     - [x] error cases (no matches for names and versions) 
  - [x] Remove-Module -ModuleInfo
  - [x] Remove-Module with constant module
      - [x] with -Force
      - [x] without -Force
  - [x] Remove-Module with readonly module
      - [x] with -Force
      - [x] without -Force
          - [ ] with constant engine module (I don't actually know how to hit this code path since these modules aren't visible in the first place via Get-Module)
          - [x] without constant engine module
  - [x] Remove-Module with module providing current session drive
  - [x] Remove-Module containing nested modules
     - [x] simple dependency A -> B
     - [x] circular dependency A -> A
  - [x] Remove-Module required by other modules which are also imported
     - [x] A -> B : remove B then A
     - [x] A -> B : remove A then B

## PR Context

This PR is aimed at increasing testing coverage for the module related cmdlets which we would like to refactor down the line.
This issue tracks the refactoring work : https://github.com/PowerShell/PowerShell/issues/8936

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

**Comment by pougetat** [comment]
Not too sure why this is is failing for Remove-Module -Name tests when eveything works locally : 

````powershell
    BeforeEach {
        Import-Module -Name "$testdrive\Modules\Foo\Foo.psd1" -Force
        Import-Module -Name "$testdrive\Modules\Bar\Bar.psd1" -Force
        Import-Module -Name "$testdrive\Modules\Baz\Baz.psd1" -Force

        (Get-Module -Name "Bar", "Baz", "Foo").Name | Should -BeExactly "Bar", "Baz", "Foo"
    }
````

With the following error : 
````
Expected exactly @('Bar', 'Baz', 'Foo'), but got @('Bar', 'Baz', 'Foo', 'Foo').
53:         (Get-Module -Name "Bar", "Baz", "Foo").Name | Should -BeExactly "Bar", "Baz", "Foo"
````

**Comment by pougetat** [comment]
Fixed the CI tests though I'm still not sure why I had two modules named "Foo" in my environment.
My initial thought was that it could be another Pester test that did not clean up its imported modules thus leaving a Foo module present but perhaps Pester tests are sandboxed in some way and these kinds of race conditions aren't possible.
Thoughts ?

File: test/powershell/Modules/Microsoft.PowerShell.Core/Remove-Module.Tests.ps1 [hunk 1]
```diff
@@ -1,5 +1,200 @@
...
+        )
+    }
+
+    BeforeEach {
+        Import-Module -Name "$testdrive\Modules\Foo\1.0\Foo.psd1" -Force
+        Import-Module -Name "$testdrive\Modules\Foo\2.0\Foo.psd1" -Force
+        Import-Module -Name "$testdrive\Modules\Bar\Bar.psd1" -Force
+        Import-Module -Name "$testdrive\Modules\Baz\Baz.psd1" -Force
+
+        (Get-Module -Name "Bar", "Baz", "Foo").Name | Should -BeExactly "Bar", "Baz", "Foo", "Foo"
```
[the remark this answers is not in the document]
Comment (pougetat) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Remove-Module.Tests.ps1]: Addressed the second comment
In regards to the first comment I can confirm that -BeExactly does indeed do what we expect to, at least in the sense that it makes sure that the list of modules we try to get are indeed present (both lists are equal).

**Review by rjmholt — APPROVED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/Remove-Module.Tests.ps1 [hunk 2]
```diff
@@ -1,19 +1,406 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
+Describe "Remove-Module -Name | -FullyQualifiedName | -ModuleInfo" -Tags "CI" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Remove-Module.Tests.ps1]: Please add new line after the license text.

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

File: test/powershell/Modules/Microsoft.PowerShell.Core/Remove-Module.Tests.ps1 (hunk 2 above)
Comment (pougetat) [reply test/powershell/Modules/Microsoft.PowerShell.Core/Remove-Module.Tests.ps1]: Done :)

**Review by adityapatwardhan — APPROVED** [review]

**Comment by iSazonov** [comment]
@pougetat Thanks for your contribution!

**Comment by fmichaleczek** [comment]
@pougetat great job ! The only missing scenario is when you import module with using [scripting.Classes.using.tests.ps1#L504](https://github.com/pougetat/PowerShell/blob/0f147ad4cc8e1a94cfe4d55cfb3d9a44a34fc3b9/test/powershell/Language/Classes/scripting.Classes.using.tests.ps1#L504) )

**Comment by rjmholt** [comment]
@fmichaleczek Please feel free to contribute more tests

**Comment by fmichaleczek** [comment]
@rjmholt sure ! if i write a test that failed (bug in powershell), what I should do with this test ?

**Comment by rjmholt** [comment]
The ideal I think is to open a PR with the test marked as `-Pending`, so that when the behaviour changes the test passes.

**Pull request merged**